### DISCUSSION TITLE
Ability to override Agent download

### DISCUSSION THREAD

**Discussion by TehNeon** [jeremylvln/Shulker#477] [Ideas]
It would be nice to have some way to either disable the download of the agent or a way to change the source of the download. The reasoning I propose this is that I'm building my own server images that have plugins baked into it (and proxy/paper jars) to prevent having to downloading assets from external sources. I'm also struggling from either my ISP or the repository host throttling my connection on downloading the agent causing it to take around 5 minutes just to finish the init-fs container on pod initialization.

**Comment by jeremylvln** [comment]
That's a good idea! I could totally add an env var that, if defined, would skip downloading the agent. I'm not 100% sure about having a configuration flag in the CRD directly (because the agent is necessary and skipping the download of it is for advanced user only). It's easy to implement, I will try to do it shortly :)

**Reply by jeremylvln** [reply]
Hi @TehNeon, I added a `skipAgentDownload` config option in #479 to both the server and the proxy. It will skip the download of the agent only.

I would love to have some feedback about your own image build. How did you it? The purpose, etc... I'm genuinely curious and at one point I will write some documentation to explain how to do it, so having your feedbacks would be awesome!

**Reply by TehNeon** [reply]
Hi @jeremylvln. Thank you for the implementation of this feature. 

The general purpose as to why I'm building my own image is that all of our plugins are custom and assembling them into our images is simplifying our CI/CD pipeline. In terms of building the image, we have a base image that pulls from`itzg/minecraft-server`.  Our image does a few changes to some of the existing environment variables that the minecraft-server image provides. As well as copies our plugins into the image.

**Reply by jeremylvln** [reply]
That's really great if you have a CI/CD pipeline for building these images! How do you manage the download of Shulker's agent plugin? You hardcoded a specific version matching the deployed Shulker version? Will this not be tedious to maintain?

**Reply by TehNeon** [reply]
As of right now I just intend on having the Shulker agent manually updated, although tedious it allows us to have it match the exact deployment of the Shulker operator we have. We're managing our cluster with a GitOps approach with ArgoCD currently, which would allow us to easily rollback changes on updates to any component such as new updates to Shulker or other components causing problems. Granted we haven't gotten to the point of having a production or release environment quite yet, I'd imagine being able to fallback onto a known working version will be handy. 

One thing I might end up looking into when our project matures more is have some sort of workflow that will pull the latest Shulker operator & agents into our images/deployment.

**Reply by jeremylvln** [reply]
I also manages my environment with ArgoCD but because I've a mirror of Shulker's Maven inside my cluster I just have overriden the repository URL Shulker uses so there are still "runtime downloads" of the assets but it's still local only. It's nice to know about other's deployments it allows to challenge our own infrastructure. Thanks for sharing.

Do not hesitate to open discussions/issues or contact me directly in private if you encounter any issue/challenge. I would gladly help you deploy Shulker to production.

I'm really looking for feedback and testimonials to help me precise the roadmap of the project!

**Reply by TehNeon** [reply]
Oh I do like the way you're handling it in your own cluster. I might end up trying to replicate that behavior or recreate something similar. 

I'm not sure if you saw my comment on the issue you opened for this feature, but it appears that update has broken pod initialization completely. (https://github.com/jeremylvln/Shulker/issues/478#issuecomment-2015542471). I probably should've just made this a new issue, but I decided to just put it on the issue that added the feature.
